adserver/models.py

```diff
@@ -941,15 +974,25 @@ def offer_ad(self, request, publisher, ad_type_slug, div_id, keywords):
             "click-proxy", kwargs={"advertisement_id": self.pk, "nonce": nonce}
         )
 
-        # This required unescaping HTML entities that bleach escapes,
-        # allowing it to be used outside of HTML contexts.
-        # https://github.com/mozilla/bleach/issues/192
-        body = html.unescape(bleach.clean(self.text, tags=[], strip=True))
+        text = self.text
+        headline = self.headline
+        body = self.body
+        cta = self.cta
+
+        if not body:
+            # This required unescaping HTML entities that bleach escapes,
+            # allowing it to be used outside of HTML contexts.
+            # https://github.com/mozilla/bleach/issues/192
+            body = html.unescape(bleach.clean(self.text, tags=[], strip=True))
+        if not text:
+            text = self.render_links(click_url)
 
         return {
             "id": self.slug,
-            "text": self.text,
+            "text": text,
+            "headline": headline,
             "body": body,
```

Comment:
Maybe we migrate the new ad data into a dict?:

```
ad:
  - headline
  - body
  - cta
```
Then keep the existing top-level body as:

```
body: bleach.clean(headline + body + cta)
```

Or maybe we rename `body` in the new ad format, so we don't have confusing overloaded terms. Perhaps `headline, content, cta`?

Reply:
I created a new dict named `copy` and put the 3 values in there.